spec/unit/parser/resource_spec.rb

```diff
@@ -283,44 +284,69 @@ class { alpha: stage => before }
       expect(edges).to include(['One[a]', 'Notify[bob says hello]'])
       expect(edges).to include(['One[b]', 'Notify[bill says hello]'])
     end
+
+    it 'should override default values with new defaults' do
+      Puppet[:code] = <<-PUPPET.unindent
+        class foo {
+          File {
+            ensure => file,
+            mode   => '644',
+            owner  => 'root',
+            group  => 'root',
+          }
+        
+          file { '/tmp/foo':
+            ensure  => directory
+          }
+        
+          File['/tmp/foo'] { mode => '0755' }
+        }
+        include foo
+        PUPPET
+
+      catalog = Puppet::Parser::Compiler.compile(Puppet::Node.new 'anyone')
+      file = catalog.resource('File[/tmp/foo]')
+      expect(file).to be_a(Puppet::Resource)
+      expect(file['mode']).to eql('0755')
+    end
   end
 
   describe 'when evaluating resource defaults' do
-    before do
-      @resource = Puppet::Parser::Resource.new("file", "whatever", :scope => @scope, :source => @source)
-    end
+    let(:resource) { Puppet::Parser::Resource.new('file', 'whatever', :scope => @scope, :source => @source) }
+
+    it 'should add all defaults available from the scope' do
+      @scope.expects(:lookupdefaults).with('File').returns(:owner => param(:owner, 'default', @source))
 
-    it "should add all defaults available from the scope" do
-      @resource.scope.expects(:lookupdefaults).with(@resource.type).returns(:owner => param(:owner, "default", @resource.source))
-      @resource.add_defaults
+      expect(resource[:owner]).to eq('default')
+    end
 
-      expect(@resource[:owner]).to eq("default")
+    it 'should not replace existing parameters with defaults' do
+      @scope.expects(:lookupdefaults).with('File').returns(:owner => param(:owner, 'replaced', @source))
+      r = Puppet::Parser::Resource.new('file', 'whatever', :scope => @scope, :source => @source, :parameters => [ param(:owner, 'oldvalue', @source) ])
+      expect(r[:owner]).to eq('oldvalue')
     end
 
-    it "should not replace existing parameters with defaults" do
-      @resource.set_parameter :owner, "oldvalue"
-      @resource.scope.expects(:lookupdefaults).with(@resource.type).returns(:owner => :replaced)
-      @resource.add_defaults
+    it 'should override defaults with new parameters' do
+      @scope.expects(:lookupdefaults).with('File').returns(:owner => param(:owner, 'replaced', @source))
 
-      expect(@resource[:owner]).to eq("oldvalue")
+      resource.set_parameter(:owner, 'newvalue')
+      expect(resource[:owner]).to eq('newvalue')
     end
 
-    it "should add a copy of each default, rather than the actual default parameter instance" do
-      newparam = param(:owner, "default", @resource.source)
+    it 'should add a copy of each default, rather than the actual default parameter instance' do
```

Comment:
Looking closer at this, I think that all instances of `Puppet::Parser::Resource::Param` should be frozen. The only place where it ever mutates seems to be when sensitive data values are unwrapped. That could just as well happen by replacing the Param instance with a copy. I think that deserves a separate ticket though.